File: dsa/concepts/13_graphs/python/topo_sort_topological_sorting.py
from collections import deque
from typing import List


"""
Problem:
Topological Sorting of a Directed Acyclic Graph (DAG)

Idea (DFS based):
In a DAG, for every directed edge (u -> v), u must appear before v.
We perform DFS and push the node into a stack *after* visiting all neighbors.
Finally, reversing the stack gives topological order.

Time:  O(V + E)
Space: O(V)
"""
def topo_sort_by_dfs(adj: List[List[int]]) -> List[int]:
    n = len(adj)
    visited = [False] * n
    stack = []

    def dfs(node: int) -> None:
        visited[node] = True
        for nei in adj[node]:
            if not visited[nei]:
                dfs(nei)
        stack.append(node)

    for i in range(n):
        if not visited[i]:
            dfs(i)

    return stack[::-1]   # reverse to get topo order


"""
Problem:
Topological Sorting using Kahn's Algorithm (BFS)

Idea:
1. Compute indegree of all nodes
2. Push all nodes with indegree 0 into queue
3. Pop node, add to topo list, reduce indegree of neighbors
4. If indegree becomes 0, push neighbor into queue

Time:  O(V + E)
Space: O(V)
"""
def topo_sort_by_kahns_algo(n: int, adj: List[List[int]]) -> List[int]:
    indegree = [0] * n

    for u in range(n):
        for v in adj[u]:
            indegree[v] += 1

    q = deque()
    for i in range(n):
        if indegree[i] == 0:
            q.append(i)

    topo = []
    while q:
        node = q.popleft()
        topo.append(node)
        for nei in adj[node]:
            indegree[nei] -= 1
            if indegree[nei] == 0:
                q.append(nei)

    return topo


# ----------- test (same as Java main) -----------
if __name__ == "__main__":
    adj = [[] for _ in range(6)]
    adj[5].extend([2, 0])
    adj[4].extend([0, 1])
    adj[2].append(3)
    adj[3].append(1)

    print("DFS Topo Sort:", topo_sort_by_dfs(adj))
    print("Kahn's Topo Sort:", topo_sort_by_kahns_algo(6, adj))

File: dsa/patterns/13_graphs/java/TopoSort_TopologicalSorting.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Stack;

public class TopoSort_TopologicalSorting {

	public static void main(String[] args) {
		TopoSort_TopologicalSorting obj = new TopoSort_TopologicalSorting();
		List<List<Integer>> adj = new ArrayList<>();
		for (int i = 0; i < 6; i++)
			adj.add(new ArrayList<>());
		adj.get(5).add(2);
		adj.get(5).add(0);
		adj.get(4).add(0);
		adj.get(4).add(1);
		adj.get(2).add(3);
		adj.get(3).add(1);

		Stack<Integer> stack = obj.topoSortByDfs(adj);
		Collections.reverse(stack);
		System.out.println(stack);

		System.out.println(obj.topoSortByKahnsAlgo(6, adj));
	}

	// Problem:
	// Idea: In DAG, for edge (u, v), in topological sorting u comes first then
	// v.
	// Always starts from vertex whose indegree is 0
	// time O(V+E)
	// space O(V)
	public Stack<Integer> topoSortByDfs(List<List<Integer>> adj) {
		int n = adj.size();

		Stack<Integer> stack = new Stack<>();
		int[] vis = new int[n];
		for (int i=0; i<n; i++) {
			if(vis[i] == 0)
				dfs(adj, i, vis, stack);
		}
		return stack;
	}

	private void dfs(List<List<Integer>> adj, int node, int[] vis, Stack<Integer> stack) {
		vis[node] = 1;
		for (int neigh : adj.get(node)) {
			if (vis[neigh] == 0)
				dfs(adj, neigh, vis, stack);
		}
		stack.add(node);
	}


	// Kahn's Algorithm:
	// 1. Insert all nodes with ingree 0 to q
	// 2. reduce the indegree of all their adj node by 1
	// 3. If indegree of adj node become 0, add it to q.
	// BFS
	public List<Integer> topoSortByKahnsAlgo(int n, List<List<Integer>> adj){
		int[] indegree = new int[n];
		for(int i=0; i<n; i++){	// find indegree
			for(int neigh: adj.get(i)){
				indegree[neigh]++;
			}
		}

		Queue<Integer> q = new LinkedList<>();	// load all indegree=0
		for(int i=0; i<n; i++){
			if(indegree[i] == 0){
				q.add(i);
			}
		}

		List<Integer> topoSort = new ArrayList<>();
		while(!q.isEmpty()){
			Integer node = q.remove();
			topoSort.add(node);
			for(int nei: adj.get(node)){
				indegree[nei]--;
				if(indegree[nei] == 0) 	// indegree arr acts as visited array
					q.add(nei);
			}
		}
		return topoSort;
	}
}
